### PR TITLE
Tiling bug fix

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #168013
* __->__ #167771

Fix for https://github.com/pytorch/pytorch/issues/166653.

Two fixes:
- We were inducing a split for broadcasted loads. e.g. (x // 16). While a split of 16 here will make the load coalesced in one of the tile vars, since the load is already in cache it's not worth splitting. And it would make the other tile var load from memory that isnt in cache. 
- Add a slight term for uncoalesced memory. This prevents doing tiling for loads which are a small % of the overall kernel.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

### CODE REVIEW

File: torch/_inductor/tiling_utils.py [hunk 1]
```diff
@@ -567,9 +599,12 @@ def remove_identity(expr: sympy.Expr) -> sympy.Expr:
...
 
 
-def get_score(addr: sympy.Expr, var_ranges: dict[sympy.Symbol, int]) -> int:
+def get_score(
+    addr: sympy.Expr, var_ranges: dict[sympy.Symbol, int], buf_names: OrderedSet[str]
+) -> int:
     """
-    Score addr according to its approximate size
+    Score addr according to its approximate size.
+    Constrained by the minimum buffer numel since the score can't be larger than the buffer.
```
Comment (v0i0) [inline torch/_inductor/tiling_utils.py]: hmm you're passing in buffer info and add the comment, but it doesn't look like this actually happens?

File: torch/_inductor/tiling_utils.py [hunk 2]
```diff
@@ -656,28 +703,40 @@ def analyze_memory_coalescing(
...
         maybe_coalesced_var = find_coalesced_var(memory_expr, var_ranges)
+        # while broadcasting vars are not technically coalesced,
+        # accesses at least stay in cache, so they provide most of the benefit.
+        # treat the same for now.
+        if maybe_coalesced_var is None:
+            maybe_coalesced_var = find_broadcast_var(memory_expr, var_ranges)
 
-        byte_multipler = 0
+        total_score = 0
         for buf_name in buf_names:
```
Comment (v0i0) [inline torch/_inductor/tiling_utils.py:719]: more for me to understand, but why would there be more than one buffer per access?

**Review by v0i0 — APPROVED** [review]

File: torch/_inductor/tiling_utils.py (hunk 1 above)
Comment (eellison) [reply torch/_inductor/tiling_utils.py]: ah thanks, that was from an old version of the code.

File: torch/_inductor/tiling_utils.py (hunk 2 above)
Comment (eellison) [reply torch/_inductor/tiling_utils.py:719]: this is just format of normalized read writes: https://github.com/pytorch/pytorch/blob/a1db54fa5ab1cc645e9d0fe192f48d00dd62711a/torch/_inductor/tiling_utils.py#L218-L227

it contains a mapping of sympy memory expr -> all buffers with that expression

File: torch/_inductor/tiling_utils.py [hunk 3]
```diff
@@ -144,6 +144,38 @@ def indexing_div_rep(
     return None
 
 
+def find_broadcast_var(
+    index: sympy.Expr, var_ranges: dict[sympy.Expr, int]
+) -> Optional[sympy.Expr]:
+    """
+    Try to find the variable that this index is broadcast over.
+    A broadcast pattern is one where consecutive values of a variable
+    access the same memory location (e.g., x // 10).
```
Comment (shunting314) [inline torch/_inductor/tiling_utils.py:153]: In general `x % 10` can also be a broadcast?

`x %10` v.s. `x // 10` just picks different dimension

Comment (eellison) [reply torch/_inductor/tiling_utils.py:153]: In this case, x % 10 will be read as coalesced so should still work the same

Comment (shunting314) [reply torch/_inductor/tiling_utils.py:153]: yea, but for `stride * (x % 10)`, it's not coalesced

Comment (eellison) [reply torch/_inductor/tiling_utils.py:153]: That won't be considered coalesced. see https://github.com/pytorch/pytorch/blob/cfe04255dee6974f37ed15a79dd96e1c3ffe3fb1/torch/_inductor/tiling_utils.py#L179

Comment (shunting314) [reply torch/_inductor/tiling_utils.py:153]: I think the main thing confuses me is the code treats:

`stride * (x % 10)` and `stride * (x // 10)` differently, while they are both broadcasting.

Comment (eellison) [reply torch/_inductor/tiling_utils.py:153]: `x % 2048` is not broadcasting. it's only with a very small modulo that it is broadcasting. in this case we're treating both coalesced and broadcasting the same, so it shouldn't matter though.

Comment (shunting314) [reply torch/_inductor/tiling_utils.py:153]:
```
import torch

@torch.compile
def f(x, y):
    return x[::2, None] + y[None, ::4]

x = torch.randn(1024, device="cuda")
y = torch.randn(2048, device="cuda")
f(x, y)
```

generates:
```
@triton.jit
def triton_poi_fused_add_slice_unsqueeze_0(in_ptr0, in_ptr1, out_ptr0, xnumel, XBLOCK : tl.constexpr):
    xnumel = 262144
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = tl.full([XBLOCK], True, tl.int1)[:]
    x1 = xindex // 512
    x0 = (xindex % 512)
    x2 = xindex
    tmp0 = tl.load(in_ptr0 + (2*x1), None, eviction_policy='evict_last')
    tmp1 = tl.load(in_ptr1 + (4*x0), None, eviction_policy='evict_last')
    tmp2 = tmp0 + tmp1
    tl.store(out_ptr0 + (x2), tmp2, None)
 ```
 
 Althoughly the generated code replace `xindex // 512` with x1 and `xindex % 512` with x0

Comment (shunting314) [reply torch/_inductor/tiling_utils.py:153]: maybe a more complex example can trigger the case that `xindex // 512` and `xindex % 512` shows up in the memory address expression directly. But the tiny example above already shows the idea
